python/ambassador/ir/irhealthchecks.py

```diff
@@ -0,0 +1,235 @@
+
+from ..config import Config
+from .irresource import IRResource
+
+if TYPE_CHECKING:
+    from .ir import IR  # pragma: no cover
+
+
+class IRHealthChecks(IRResource):
+
+    # The list of mappers that will make up the final health checking config
+    _mappers: Optional[List[Dict[str, Union[str, int, Dict]]]]
+
+    # The IR config, used as input from a `health_checks` field on a Mapping
+    _ir_config: List[Dict[str, Union[str, int, Dict]]]
+
+    def __init__(
+        self,
+        ir: "IR",
+        aconf: Config,
+        health_checks_config: List[Dict[str, Union[str, int, Dict]]],
+        rkey: str = "ir.health_checks",
+        kind: str = "IRHealthChecks",
+        name: str = "health_checks",
+        **kwargs,
+    ) -> None:
+        self._ir_config = health_checks_config
+        self._mappers = None
+        super().__init__(ir=ir, aconf=aconf, rkey=rkey, kind=kind, name=name, **kwargs)
+
+    # Return the final config, or None if there isn't any, either because
+    # there was no input config, or none of the input config was valid.
+    def config(self) -> Optional[List[Dict[str, Union[str, int, Dict]]]]:
+        if not self._mappers:
+            return None
+        return self._mappers
+
+    def setup(self, ir: "IR", aconf: Config) -> bool:
+        self._setup(ir, aconf)
+        return True
+
+    def _setup(self, ir: "IR", aconf: Config):
+        # Dont post any errors if there is empty config
+        if not self._ir_config:
+            return
+
+        # The health checking config must be an array
+        if not isinstance(self._ir_config, list):
+            self.post_error(
+                f"IRHealthChecks: health_checks: field must be an array, got {type(self._ir_config)}, Ignoring...",
+                log_level=logging.ERROR,
+            )
+            return
+
+        # Do nothing (and post no errors) if there's config, but it's empty.
+        if len(self._ir_config) == 0:
+            return
+
+        # If we have some configuration to deal with, try to load it, and post any errors
+        # that we find along the way. Internally, _load_config will skip any health checks
+        # that are invalid, preserving other rules. This prevents one bad health check from eliminating
+        # the others.
+        self._mappers = self._generate_mappers()
+        if self._mappers is not None:
+            ir.logger.debug("IRHealthChecks: loaded mappers %s" % repr(self._mappers))
+
+    def _generate_mappers(self) -> Optional[List[Dict[str, Union[str, int, Dict]]]]:
+        all_mappers: List[Dict[str, Union[str, int, Dict]]] = []
+
+        # Make sure each health check in the list has config for either a grpc health check or an http health check
+        for hc in self._ir_config:
+
+            grpc_health_check = hc.get("grpc_health_check", None)
+            http_health_check = hc.get("http_health_check", None)
+
+            if grpc_health_check is None and http_health_check is None:
```

Comment:
We can't move this logic earlier in the CRD Validation, correct? The only reason we can do that on the Edge Stack side is because we explicitly call our own Validation func?

I know OpenAPI v3 supported it but I think OpenAPIv2 doesn't which is still used internally by K8s, correct?

Reply:
The reason I ask is because it would be nice from a user perspective if this type of logic was at the apply time and rejected before ever getting persisted.

Reply:
We can for fields where one is required all the time, but unfortunately there was no easy way to check for a case when neither field is required but at least one must be present. I was really hoping there would be a way to get that check in earlier so that you would be prevented from applying the CRD :cry:

Reply:
Ok that is what I thought, we do not call any additional validation functions in the watch phase prior to passing it along to `diagd`. As you point out ideally it would be handled via openAPI 3's support for oneOf but kubernetes doesn't support it yet due to converting it to openAPI v2.

I will keep this in mind when thinking about ways we can improve this story overall.

Reply:
I'm confused why this still needs to be here with the

```go
// +kubebuilder:validation:MinProperties=1
// +kubebuilder:validation:MaxProperties=1
type HealthCheckLocation struct {
```

validations.